feat: add registry1 zarf flavors to uds core

zachariahmiller opened the pull request:
## Description
Adds zarf flavors for upstream images and IB. Includes refactoring of zarf packages using composability to cleanup the config and keep things dry and refactoring of the values files to image flavors specific + common. 

Both flavors deploy successfully on their respective architectures. On arm64 (or at least darwin arm64) when using the registry1(x86) flavor it fails at metrics server regarding the istio init container, which seems to be an issue [emulating x86 on mac](https://github.com/istio/istio/issues/36762#issuecomment-1034070488). 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [x] Test, docs, adr added or updated as needed
- [ ] [Contributor Guide Steps](https://github.com/defenseunicorns/uds-template-capability/blob/main/CONTRIBUTING.md)(https://github.com/defenseunicorns/uds-template-capability/blob/main/CONTRIBUTING.md#submitting-a-pull-request) followed

zachariahmiller committed:
add flavors

zachariahmiller committed:
fix promtail chart image tag ref and make validate-all standalone

jeff-mccoy committed:
playing with composition

jeff-mccoy committed:
composition ignores required

zachariahmiller committed:
partial flavors composition refactor

zachariahmiller committed:
Update grafana to new pattern/package structure

zachariahmiller committed:
refactor flavors with common component pattern

zachariahmiller committed:
refactor flavors with common component pattern

zachariahmiller commented on src/promtail/values/registry1-values.yaml:
```diff
@@ -0,0 +1,165 @@
+            source_labels:
+            - __meta_kubernetes_pod_name
+            target_label: pod
+          - action: replace
```
@rjferguson21 pointed out i missed this. Deleting the redundant values when i update the merge conflicts

zachariahmiller commented on src/neuvector/values/registry1-values.yaml:
```diff
@@ -0,0 +1,108 @@
+        client_id: null
+        client_secret: null
+        default_role: null
+        enable: true
```
missed cleanup here too. resolving alongside the merge conflict.

zachariahmiller committed:
merge main into flavors, resolve conflicts, fix missed values deletions

zachariahmiller committed:
Update CI, remove missed, extraneous, values in flavor specific values files, fix core-istio create task

zachariahmiller committed:
fix missed merge conflict in settings.json

zachariahmiller committed:
run pepr format

zachariahmiller committed:
fix missing security context in neuvector monitor values

zachariahmiller committed:
update CI to add running (only) a full test for registry1 flavor

zachariahmiller committed:
add registry1 login

mjnagel commented on .github/workflows/test.yaml:
```diff
@@ -39,13 +42,20 @@ jobs:
       - name: Test a single source package
-        if: ${{ inputs.package != 'all' }}
-        run: uds run test-single-package
+        if: ${{ inputs.package != 'all' && matrix.flavor == 'upstream'}}
```
Don't have a _strong_ opinion here but I wonder if it's more valuable to switch this to use registry1 for the single package tests (instead of upstream). My thinking here is that people are more likely to be doing upstream tests locally (and those are more likely to "just work" since its all upstream), so those pieces are generally going to be tested before being pushed. With that in mind...short circuiting CI on registry1 might make more sense?

zachariahmiller replied:
I do agree we are more likely to catch an issue on the registry1 tests. What is in here now is what we had agreed upon previously. 
I'm inclined to possibly just do both since the single capability tests are on public (free) runners and will still provide the short circuit when needed, with slightly more feedback, slightly less CI logic and no expense.

zachariahmiller replied:
based on how this ci testing has been going im leaning towards doing both sets of single capability tests...

zachariahmiller replied:
we're doing both after the workflow refactor.

mjnagel commented on .yamllint:
```diff
@@ -0,0 +1,38 @@
+ignore:
+  - 'k3d/local/manifests/metallb/metallb-native.yaml'
+  - '**/.terraform/**'
+  - '**/chart/templates**'
```
```suggestion
  - '**/chart/templates**'
  - 'node_modules/**'
```
I had a bunch of bogus linting issues locally on the `node_modules` folder.

mjnagel commented on src/istio/zarf.yaml:
```diff
@@ -1,7 +1,7 @@
 kind: ZarfPackageConfig
```
Let me know what you think about these changes - https://github.com/defenseunicorns/uds-core/pull/108 - I think this feels closer to the common pattern with other src packages?

zachariahmiller replied:
merged. I agree with that pattern i had started going down that road and then got to the gateways which didnt fit and figure i'd try to get feedback first.

mjnagel commented on tasks/create.yaml:
```diff
@@ -5,7 +5,7 @@ tasks:
 
       - description: "Create the UDS Core Standard Zarf Package"
-        cmd: "zarf package create packages/standard --confirm"
+        cmd: "zarf package create packages/standard --confirm  --flavor ${UDS_FLAVOR}"
```
Wondering if we should put a default flavor to simplify the dev experience (i.e. set default `UDS_FLAVOR` to `upstream` so that local dev can remain the same as previous experience.

zachariahmiller replied:
I would like a default. I'll test locally to see if setting UDS_FLAVOR as a top level variable works as expected to facilitate this and update accordingly if it does, otherwise we can get that sorted on the uds cli side and then add it.

mjnagel commented on .github/workflows/tag-and-release.yml:
```diff
@@ -44,8 +47,15 @@ jobs:
+
       - name: Publish UDS Zarf Package
-        run: uds run -f tasks/publish.yaml packages
+        run: uds run -f tasks/publish.yaml packages --set UDS_FLAVOR=${{ matrix.flavor }}
```
I think for the bundle publishing we will need something special for the naming, but might be wrong. I don't think uds-cli appends the flavor to tags yet?

zachariahmiller replied:
yeah... I wasnt sure on the state of that or planned state for the bundle publishing. Easy enough to solve for either way.

zachariahmiller replied:
I verified uds wasnt doing anything with the flavor. Since there isnt a way to override the tag on publish, I updated to make the registry path .../bundles/${FLAVOR}/

mjnagel replied:
While I don't love how long that makes the artifact name/path (ex: `ghcr.io/defenseunicorns/packages/uds/bundles/registry1/k3d-core-istio:0.8.0`) I think this is a simple and reasonable solve.

One way I was considering that might be a bit more in line with how zarf handles flavors would be to dynamically tack on the flavor to the tag/version during the publish step (`yq -i '.metadata.version = "0.8.0-registry1"' bundles/k3d-standard/uds-bundle.yaml` before build/publish). While I like the outcome of that...it feels pretty hacky 🤷. We could do the same thing "statically" by duplicating the bundle files (committing with the flavor in the version), but that's not very DRY.

All that to say...I'm good with the way you handled it. I've probably already thought too much about something that isn't meant to be consumed outside of dev 😅

zachariahmiller replied:
Yeah so I'm not a huge fan of the way i did it being different from how zarf tags it autotically, but figured it was between that, waiting for a way for UDS to handle flavors automatically or what you suggested above. I think I slightly prefer the outcome of your suggestion because it matches how zarf gets published but as you said the yq approach is hacky and the static approach feels excessive. If anyone else has strong opinions I'd welcome them to weigh in :D

zachariahmiller replied:
@UncleGedd relevant discussion on uds and flavors.

UncleGedd replied:
Appreciate the discussion @zachariahmiller @mjnagel ! Would one of you mind opening an issue in UDS CLI to continue the conversation and discuss proposed options?

mjnagel replied:
Opened https://github.com/defenseunicorns/uds-cli/issues/332 to have this discussion. Going to resolve this thread for now since I think we have a path forward and as mentioned before these bundles are pretty explicitly for dev use and changing the name down the road should not be a big deal.

mjnagel committed:
chore: refactor istio flavors (#108)

Internal PR, not to main branch.

I think this is closer to the pattern for other src packages.

zachariahmiller committed:
Update .yamllint

Co-authored-by: Micah Nagel <micah.nagel@defenseunicorns.com>

mjnagel committed:
chore: linting fixes

zachariahmiller commented:
I'll work on resolving the merge conflicts from the operator PR, testing bundle publish flavor behavior and setting a default flavor in the tasks tonight and tomorrow morning.

Intermittent neuvector deploy is still an issue as they dont actually expose security context in the monitor chart and it sometimes happens to work by continuing past it anyways. Not sure what the immediate fix for this is, but @mjnagel put in a pr to the upstream chart to fix on their side that we can use once it gets added. https://github.com/neuvector/neuvector-helm/pull/355

zachariahmiller committed:
rename UDS_FLAVOR to FLAVOR and update bundle publish path to support variants.

zachariahmiller committed:
chore: merge main into flavors and resolve conflicts

zachariahmiller committed:
fix: disable neuvector monitor chart until securitycontext is exposed in the upstream and/or service monitors are in place

zachariahmiller committed:
fix: re-add required: true on istio control plane so single capability tests can execute successfully

zachariahmiller committed:
chore: resolve yaml lint issues after merge, pepr format

zachariahmiller committed:
chore: add missing default task at top level taskfile, update upgrade to be per flavor

zachariahmiller commented:
Before being able to merge, I think we'll need to change the repo settings for expected tests from this:
![image](https://github.com/defenseunicorns/uds-core/assets/71521611/4e263f8c-b6f4-4d89-a414-98aceffd4683)

to these:
![image](https://github.com/defenseunicorns/uds-core/assets/71521611/8285f8ad-9ae3-4437-9bee-9144c368aaa8)

zachariahmiller committed:
chore: update uds cli version to properly set the FLAVOR variable for uds run

zachariahmiller committed:
chore: add regitry1 login for upgrade job

zachariahmiller committed:
chore: fix settings.json uds version, refactor test apps to zarf package

zachariahmiller committed:
fix: add neuvector peer auth exception

zachariahmiller commented:
At this point the only outstanding thing should be that loki is failing on upgrade because the gateway wont update. Deleting the old one successfully continues the upgrade to completion.
```
message: '0/1 nodes are available: 1 node(s) didn''t match pod anti-affinity rules.  
preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod
```

this might just be a single node issue. Diving into the chart to see if we can change the restart policy or something for single node.

mjnagel commented on README.md:
```diff
@@ -28,7 +28,7 @@ The core applications are:
+| [UDS CLI](https://github.com/defenseunicorns/uds-cli/releases) | 0.6.1           |
 | [NodeJS](https://nodejs.org/en/download/)                      | LTS or Current  |
 
 <!-- endtable -->
```
I can't comment on the actual lines but we'll want to update the quickstart `uds deploy` commands to reference the new bundle names with flavors.

zachariahmiller replied:
done

mjnagel commented on src/neuvector/common/zarf.yaml:
```diff
@@ -0,0 +1,35 @@
+      #   namespace: neuvector
+      #   gitPath: charts/monitor
+      #   valuesFiles:
+      #     - ../values/monitor-values.yaml
```
I'm *okay* with this but in my head we could just put something like this in the registry1 values file:

```yaml
exporter:
  enabled: false
```

Ideally less churn/changes when we can re-enable it?

zachariahmiller replied:
thats fine. I did it this way because it seemed like the most obvious/apparent way to see that it was done and why. Removing the comments and updating the values with the comment on there as a note

zachariahmiller committed:
Merge branch 'main' into 60-add-registry1-zarf-flavors-to-uds-core

zachariahmiller committed:
fix: remove deployment strategy

zachariahmiller committed:
refactor workflow matrix

zachariahmiller committed:
remove single capability conditional

zachariahmiller committed:
chore: update concurrency group name

zachariahmiller committed:
chore: refactor how we temporarily disable the neuvector exporter

zachariahmiller committed:
chore: uncomment monitor chart in neuvector common

zachariahmiller committed:
docs: update readme to reflect flavor usage and update registry paths

mjnagel committed:
ci: fix upgrade test on release please PRs (#113)

## Description

Previous logic was problematic and would fail on PRs where the version
was getting updated (i.e. release-please PRs). Using `crane`/`zarf tools
registry` to get the tags and sort/tail for latest seems like the
simplest approach.

## Related Issue

N/A

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Other (security config, docs update, etc)

## Checklist before merging

- [x] Test, docs, adr added or updated as needed
- [x] [Contributor Guide
Steps](https://github.com/defenseunicorns/uds-template-capability/blob/main/CONTRIBUTING.md)(https://github.com/defenseunicorns/uds-template-capability/blob/main/CONTRIBUTING.md#submitting-a-pull-request)
followed

zachariahmiller committed:
fix: update publish to ignore arm64 for registry1 flavor, remove extra line in neuvector yaml left over from merge conflict resolution

mjnagel reviewed (dismissed):
LGTM! Great work on making this more flavorful.

zachariahmiller commented:
@jeff-mccoy This is good to go, the only thing is the required check no longer exists and needs to be updated in the settings. 
![image](https://github.com/defenseunicorns/uds-core/assets/71521611/92e02789-4c04-4254-8e9a-2a2001bfe61c)

I'm able to make the changes and merge, but you had said you wanted to review. PTAL or let me know if you are good with the extensive review of this with @mjnagel that has already happened so i can move forward with merging this.

jeff-mccoy commented on .yamllint:
```diff
@@ -0,0 +1,39 @@
+yaml-files:
```
not familiar with this config file. can you esplain plz?

zachariahmiller replied:
I added a [lint task](https://github.com/defenseunicorns/uds-core/blob/b7a41e8acbcd9a5b717ba7d6304ec66cebd54dc5/tasks/lint.yaml#L8) that runs as yaml lint check and npx pepr format. This is the config file for the yaml lint. It is the same yam,l lint we use in [dubbd](https://github.com/defenseunicorns/uds-package-dubbd/blob/642e948ebdc67952d00ed6dfeb8088f326d9e863/.yamllint)

zachariahmiller replied:
fwiw i would have preferred this to function more like pepr format and actually do the formatting but it seemed like there were limited options for that for yaml specifically that wouldnt potentially cause more issues.

jeff-mccoy replied:
sorry I meant, what is the `yamllint` binary? I don't have it on my system or see where we're installing it

zachariahmiller replied:
right. I dont install it because we dont install the tools in our run tasks per your request earlier on in this project (like uds and zarf install steps we originally had). yamllint can be installed via `pip install yamllint`. See dubbd's ci call: https://github.com/defenseunicorns/uds-package-dubbd/blob/642e948ebdc67952d00ed6dfeb8088f326d9e863/.github/workflows/ci-on-pull-request-k3d.yaml#L29

If you'd like i can add this as a CI step. I was initially adding so i could check formatting before pushing up changes as i said previously and wasnt trying to make this PR even larger by adding it to CI.

zachariahmiller replied:
added this as a ci step for it does a pepr format --check-validate and the yaml lint check in ci. Locally the task is still there to fix the formatting with pepr.

In the future i'd be happy to evaluate a different tool than yamllint especially if it can auto format in a non-breaking way, which so far the tools i've looked at either wont fix or cause issues with the zarf package schema once formatted.

jeff-mccoy commented on package.json:
```diff
@@ -36,6 +36,7 @@
   },
   "devDependencies": {
     "@jest/globals": "29.7.0",
+    "@types/jest": "^29.5.11",
```
not a pinned dep, also why is this needed here? I don't see any other TS changes anywhere.

zachariahmiller replied:
I can delete this but i dont know where its coming from... To your point no TS changes.

jeff-mccoy commented on tasks/lint.yaml:
```diff
@@ -0,0 +1,8 @@
+    actions:
+      - description: "Pepr Format"
+        cmd: npx pepr format
+      - description: yaml lint
```
still not familiar with this tool or where it is being installed. Also will this have issue with prettier?

zachariahmiller replied:
I dont think it would have an issue with prettier afaik prettier isnt doing anything with the yaml. That being said im not really familiar with prettier aside from having it baked in to some js projects i've worked on in the past. 

Right now that is a helper task and the answer is pip. This is the linter dubbd uses in their pipeline for yaml https://github.com/defenseunicorns/uds-package-dubbd/blob/642e948ebdc67952d00ed6dfeb8088f326d9e863/.github/workflows/ci-on-pull-request-dubbd-aws.yaml#L28

I had added these because i was getting annoyed i kept having to npx pepr format when pulling in new code even though i wasnt touching pepr code in my changes and wanted to make sure the ton of yaml was all being formatted in my dev.

I think these (or similar) would be added in a separate ticket as part of the ci process.

jeff-mccoy replied:
yeah glad we have a step for it for sure

jeff-mccoy commented on tasks/deploy.yaml:
```diff
@@ -26,8 +26,12 @@ tasks:
 
   - name: latest-package-release
     actions:
+      - description: "Get latest tag version from OCI"
```
can you explain this change, seems unrelated to flavors?

zachariahmiller replied:
sure. The upgrade process got merged into main prior to me being able to get this merged. latest-package-release is a part of that change, not flavors.

mjnagel replied:
Full context on this one - https://github.com/defenseunicorns/uds-core/pull/113 - the previous logic was problematic in release please PRs.

jeff-mccoy replied:
cool thx for the clarification

jeff-mccoy commented on tasks/publish.yaml:
```diff
@@ -14,50 +17,62 @@ tasks:
+          pkgPath="ghcr.io/defenseunicorns/packages/uds/bundles/${FLAVOR}/k3d-core"
+          # dont tag arm64 for registry1 since IB images are only amd64
+          if [ "${FLAVOR}" != "registry1" ]; then 
+            zarf tools registry copy ${pkgPath}:${VERSION}-arm64 ${pkgPath}:arm64
```
I'm not totally sure we really need "latest" shortcuts for the other flavors besides upstream since these are really just meant for quick testing/demo

zachariahmiller replied:
I dont really think we need them either way. I was just updating the publish step to have parity with the what was already happening.

IMO doing upstream one way and the other a different way is going to confuse someone trying to consume.

jeff-mccoy replied:
k we can save for another time to think about that

jeff-mccoy reviewed (changes requested):
Left a few comments, but overall looks pretty good. Couple questions:

1. How will this work with Renovate with the mess of version tracking between IB and upstream?

2. Not sure we should be testing both flavors for single-app tests are those are really just circuit-breaker tests to speed up failure feedback.

jeff-mccoy commented:
these skipped tests also look pretty ugly, think we need to address that as well. 

<img width="857" alt="Screenshot 2024-01-19 at 1 11 25 AM" src="https://github.com/defenseunicorns/uds-core/assets/882485/ff80bb4d-9daf-45ce-b8aa-494319fd43d6">

zachariahmiller commented:
> these skipped tests also look pretty ugly, think we need to address that as well.
> <img alt="Screenshot 2024-01-19 at 1 11 25 AM" width="857" src="https://private-user-images.githubusercontent.com/882485/297982638-ff80bb4d-9daf-45ce-b8aa-494319fd43d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU2NzQ1ODQsIm5iZiI6MTcwNTY3NDI4NCwicGF0aCI6Ii84ODI0ODUvMjk3OTgyNjM4LWZmODBiYjRkLTlkYWYtNDVjZS1iOGFhLTQ5NDMxOWZkNDNkNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDExOVQxNDI0NDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iNDI3MWMxNjI2MjQ2NDZmYmVjMDJkNmY4OTMzNWI3YmIwMGFjM2VlZmIxYTIyYjRhOTNjMjY4Mzc1ZmQ3NDJhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.yOiWv8YkuWeikoJoxygZ019JLZej4hf3BVOy6afAoKo">

I dont disagree. Again, related to the upgrade. Not sure how to fix this without a more significant refactor of the workflows. IMO that could be addressed in a follow on PR, but whatever the team wants to do.

zachariahmiller commented:
> Left a few comments, but overall looks pretty good. Couple questions:
> 
>     1. How will this work with Renovate with the mess of version tracking between IB and upstream?
> 
>     2. Not sure we should be testing both flavors for single-app tests are those are really just circuit-breaker tests to speed up failure feedback.

1. The goal with the way I broke out the repo/image/tag pieces for each flavors values file was supposed to deal with that. When i was testing the renovate config i was testing with a WIP version of flavors. It was mostly working, one of the issues we had was that when image versions werent referenced at all (some of upstream) the renovate pr wouldnt always be consistent, so the current pattern is to make the registry1 anbd upstream image specific pieces match completely so we shouldnt have any weirdness with renovate. I think we should be in a pretty good place as far as that goes and we can always tweak the renovate config as needed. Most of them should update correctly grouped by app including both flavors. I'm sure there will be some exceptions (wouldnt be surprised if something takes an issue with this "registry.k8s.io/ingress-nginx/kube-webhook-certgen:v20221220-controller-v1.5.1-58-g787ea74b6" for example :D ) 

2. Maybe we dont, but those tests are free. There was a discussion earlier in the review on that https://github.com/defenseunicorns/uds-core/pull/63#discussion_r1454036240 i think given it is no cost to us with some potential minor gain (aside from whether we chose to test registry1 or upstream) there isnt really a compelling reason to only test one. As it stands though if we did that, which is what i originally implemented, we'd end up with the same skipped test thing you se now for the upgrades, which would require refactoring the workflow files more.

zachariahmiller committed:
chore: remove extra package.json dev dependency

mjnagel committed:
ci: refactor to avoid skipped job ugliness (#115)

Internal PR between branches, separates upgrade job to different
workflow file to avoid skipped jobs on PR display.

zachariahmiller committed:
fix: add retry action action on registry1 docker login to mitigate intermittent job failures due to registry. consolidate docker login action into setup

zachariahmiller committed:
fix: pass secrets as inputs to setup action

zachariahmiller committed:
ci: add formatting checks before pipeline execution

zachariahmiller committed:
fix: add missing node setup to format check

zachariahmiller committed:
ci: add missing deps to formatting check task

zachariahmiller committed:
chore: fix formatting per checks

zachariahmiller committed:
ci: update step name

zachariahmiller commented:
@jeff-mccoy here is an example of what we expect the renovate updates to look like more or less. This was in my fork before making the registry1 and upstream values files both have all the image refs rather than relying on it just working in upstream. With the values file changes it will update the upstream-values, registry1-values, and the zarf.yaml images and chart versions in one PR.
![image](https://github.com/defenseunicorns/uds-core/assets/71521611/c25e78bd-a942-4081-9cfc-0230147640d3)

zachariahmiller commented:
For clarity/transparency we were trying to determine why the fast failure started only working on a click ops cancel and not a job failure in the updated ci. 

It looks like we have two reasonable options to get it back to that functionality.

Option 1: 
Use something like this composite action but built on bash/sh not powershell https://github.com/Lombiq/GitHub-Actions/blob/dev/.github/actions/cancel-workflow/action.yml#L27 This should work but it looks like from @mjnagel research there are some downsides: 

> This is the simplest way I've seen to cancel other jobs. But the downside of that is that it results in everything showing cancelled so it's a bit harder to find the failed job...

Option 2: Go back to the previous workflow setup (all calling just the one test workflow) while expanding the matrix. I tested/mocked this in an outdated fork as it was the easiest way to do so and it seems to work, downside is that the actual display of the job names and lack of grouping is a little less readable. The mocked up version of that can be seen here:
https://github.com/zachariahmiller/uds-core/pull/42
https://github.com/zachariahmiller/uds-core/actions/runs/7595497465/job/20688138631?pr=42

zachariahmiller commented:
@mjnagel I dug into the matrix functionality a little more. It looks like with a combination of excludes and includes in the matrix definition, relying on order they are processed per the [docs](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategymatrixexclude) seems to work. For example:

```
    strategy:
      matrix:
        package: ${{ fromJSON(needs.check-paths.outputs.packages) }}
        flavor: [upstream, registry1]
        test_type: [install, upgrade]
        exclude:
          - test_type: upgrade
        include:
          - test_type: upgrade
            flavor: registry1
            package: all
          - test_type: upgrade
            flavor: upstream
            package: all
```

Excludes the single upgrade tests and still does both for all while maintaining the fail fast functionality. 

As you can see in the screenshot, the single upgrades were excluded, install and upgrade test ran for "all" and the failed job short circuited the all the sibling workflows spawned from the matrix config while still failing the pipeline.
![image](https://github.com/defenseunicorns/uds-core/assets/71521611/5eb87b9f-fe75-41b3-94a4-d95909a5bb6b)
![image](https://github.com/defenseunicorns/uds-core/assets/71521611/fb1dd83a-8ef7-4c4e-973e-1e21f9fdaa64)

mjnagel commented:
@zachariahmiller I think that makes sense. This slight modification (if it works) would make it a bit simpler I think?

```
    strategy:
      matrix:
        package: ${{ fromJSON(needs.check-paths.outputs.packages) }}
        flavor: [upstream, registry1]
        test_type: [install]
        include:
          - test_type: upgrade
            flavor: registry1
            package: all
          - test_type: upgrade
            flavor: upstream
            package: all
```

If I read the docs right I think that'll work? May even be able to drop the flavor from the include but feel less confident on that.

zachariahmiller commented:
If that works, agreed. As far as dropping flavor, I don't know if it will... With the combo of include and exclude the flavor was required. IIRC exclude worked on partial match, include didn't.

jeff-mccoy reviewed (approved):
🚀

jeff-mccoy commented:
Branch protection updated, g2g.

<img width="823" alt="Screenshot 2024-01-21 at 3 05 06 AM" src="https://github.com/defenseunicorns/uds-core/assets/882485/a70377b0-83ec-402e-875a-14e8f5b16370">
